src/common/hooks/useAuth/index.tsx

```diff
@@ -0,0 +1,92 @@
+import { roleAtom, setTokenAtom, tokenAtom, UserRole } from 'atoms/authAtom';
+import { useAtom } from 'jotai';
+
+import { Navigate, useNavigate } from "react-router-dom";
+import { AuthResult } from './types';
+import { useAuthLogin, useAuthLogout } from 'queries/modules/auth/useAuthQuery';
+import { removeTokenFromLocalStorage } from 'utils/authStorage';
+import { LoginRequest } from 'api/modules/auth/types';
+
+/**
+ *
+ * @param requiredRoles // 관리자 Role
+ * @param redirectPath // 리다이렉트 경로
+ * @returns {AuthResult}
+ * @description
+ * 인증 관련 Hook
+ * - requiredRoles: 해당 페이지에 접근하기 위해 필요한 Role
+ * - redirectPath: 인증되지 않은 사용자가 접근 시 리다이렉트 될 경로
+ * - isAuthenticated: 인증된 사용자인지 여부
+ * - AuthGuard: 인증된 사용자만 접근 가능한 페이지를 위한 컴포넌트
+ * - login: 로그인 함수
+ * - logout: 로그아웃 함수
+ */
+
+const useAuth = (requiredRoles: UserRole[], redirectPath: string = "/sign-in"): AuthResult => {
+  const [role] = useAtom(roleAtom);
+  const [token] = useAtom(tokenAtom);
+  const setToken = useAtom(setTokenAtom)[1];  // 쓰기 전용 토큰 설정 함수 사용
+
+  const navigate = useNavigate();
+
+  // 로그인 함수
+  const { mutate: loginMutate, error: loginError } = useAuthLogin({
+    onSuccess: (data) => {
+      // 로그인 성공 시 토큰과 역할 저장
+      const { accessToken, refreshToken } = data.response
+
+      setToken({ accessToken, refreshToken }); // 토큰 저장
+      navigate('/'); // 로그인 후 메인 페이지로 이동
+    },
+  });
+
+  // 로그아웃 함수
+  const { mutate: logoutMutate } = useAuthLogout({
+    onSuccess: () => {
+      // 로그아웃 성공 시 토큰과 역할 초기화
+      removeTokenFromLocalStorage();
+      setToken({ accessToken: null, refreshToken: null });
+      navigate('/sign-in'); // 로그아웃 후 로그인 페이지로 이동
+      // setRole(role);
+    },
+  });
+
+  /**
+   * TODO: 리프레시 토큰 구현
+   * 리프레시 토큰은 백엔드 제공 후 axios interceptor를 통해 자동 갱신 구현해야함
+   */
+  // const { mutate: refreshMutate } = useAuthRefresh({
+  //   onSuccess: (data) => {
+  //     // 토큰 갱신 성공 시 토큰 업데이트
+  //     setToken({ accessToken: data.response.accessToken, refreshToken: token.refreshToken });
+  //   },
+  // });
+
+  // 인증된 사용자인지 확인
+  const isAuthenticated = !!token.accessToken && requiredRoles.includes(role);
+
+  // AuthGurad
+  const AuthGuard = ({ children }: { children: JSX.Element }) => {
+    if (!token) {
+      return <Navigate to={redirectPath} state={{ from: location }} replace />; // 로그인 안된 경우 리디렉트
+    }
+
+    if (!isAuthenticated) {
+      return <Navigate to={redirectPath} replace />;
+    }
+    return children;
+  };
+
+    // 로그인 함수
+  const login = async (payload: LoginRequest) => {
+    loginMutate(payload);
+  };
+
+  const logout = () => {
+    logoutMutate({ refreshToken: token.refreshToken! });
+  }
+
+  return { isAuthenticated, AuthGuard, login, logout, error: loginError?.message ?? null, role };
```

Comment:
요렇게가 더 명확하지 않을까요? 
사용하는 쪽에서 error라고 불러오면 auth와 관련해서 언제 발생하는 에러인지 이해하기 어려울 것 같아서요!
```suggestion
  return { isAuthenticated, AuthGuard, login, logout, loginError: loginError?.message ?? null, role };
```

Reply:
생각해보니까 useQuery의 error같은 경우에는 이렇게 별도로 값을 만들어주는게 좋을 것 같은데요! 
mutation의 error같은 경우에는 해당 mutation을 실행하는 코드에서 try-catch로 표현해주는게 더 직관적일 것 같아서 요렇게 외부로 노출시켜주지 않아도 괜찮을 것 같은데 어떠신가요?

Reply:
추후에 백엔드 개발이 되고 고민해보려 했던 사항이에요. 에러메세지를 **ERROR_MSG_MAP** 같은 맵 객체를 만들어서 던져줄 생각이였는데, 우선 현재는 useQuery의 에러만 던져주고 있어서 승완님 의견에 동의합니다~~ 현재는 query를 직접 사용하는 곳에서 사용하는게 더 명시적이고 좋을거 같아요